mesonbuild/wrap/wrap.py

```diff
@@ -566,18 +569,24 @@ def get_file_internal(self, what: str) -> str:
 
             return path.as_posix()
 
+    def unpack_archive(self, path):
+        with tempfile.TemporaryDirectory() as tmpdir:
+            shutil.unpack_archive(path, tmpdir)
+            src = os.listdir(tmpdir)
+            if len(src) != 1:
+                raise WrapPatchFormatException("Error patch archive format.")
+            self.copy_tree(src[0], os.path.join(self.subdir_root, self.wrap.name))
+
     def apply_patch(self) -> None:
         if 'patch_filename' in self.wrap.values and 'patch_directory' in self.wrap.values:
             m = 'Wrap file {!r} must not have both "patch_filename" and "patch_directory"'
             raise WrapException(m.format(self.wrap.basename))
         if 'patch_filename' in self.wrap.values:
             path = self.get_file_internal('patch')
             try:
-                shutil.unpack_archive(path, self.subdir_root)
+                self.unpack_archive(path)
             except Exception:
```

Comment:
This try/except block is useless now, just call self.unpack_archive() without try.